// react-sequence-animator/dist/src/SequenceAnimator/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ease = undefined;

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _easingUtils = require('easing-utils');

var Easings = _interopRequireWildcard(_easingUtils);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function doEase(pos, start, end) {
  return start + (end - start) * pos;
}

var ease = exports.ease = function ease(easeName) {
  return function (t, start, end, duration) {
    return doEase(Easings[easeName](t / duration), start, end);
  };
};

var SequenceAnimator = (_temp = _class = function (_Component) {
  (0, _inherits3.default)(SequenceAnimator, _Component);

  function SequenceAnimator() {
    (0, _classCallCheck3.default)(this, SequenceAnimator);

    var _this = (0, _possibleConstructorReturn3.default)(this, (SequenceAnimator.__proto__ || (0, _getPrototypeOf2.default)(SequenceAnimator)).call(this));

    _this._getFrame = function () {
      var frame = _this.state.frame;
      var children = _this.props.children;

      var childrenArr = _react2.default.Children.toArray(children);

      return childrenArr.length >= frame ? childrenArr[frame] : null;
    };

    _this._playAnimation = function () {
      _this._animationFrame = requestAnimationFrame(_this._onAnimate);
    };

    _this._onAnimate = function (timestamp) {
      var _this$props = _this.props,
          onSequenceEnd = _this$props.onSequenceEnd,
          onAnimationStop = _this$props.onAnimationStop,
          children = _this$props.children,
          loop = _this$props.loop,
          easing = _this$props.easing,
          duration = _this$props.duration;

      var childrenArr = _react2.default.Children.toArray(children);

      if (!_this._animationStart) {
        _this._animationStart = timestamp;
      }

      var nextFrame = Math.floor(ease(easing)(timestamp - _this._animationStart, 0, childrenArr.length, duration));

      if (nextFrame > childrenArr.length - 1) {
        onSequenceEnd();

        if (loop) {
          nextFrame %= childrenArr.length;
          _this._animationStart = timestamp;
        } else {
          nextFrame = -1;
        }
      }

      if (nextFrame > -1) {
        _this.setState({ frame: nextFrame }, function () {
          _this._playAnimation();
        });
      } else {
        onAnimationStop();
      }
    };

    _this.state = {
      frame: 0
    };
    return _this;
  }

  (0, _createClass3.default)(SequenceAnimator, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var autoplay = this.props.autoplay;


      if (autoplay) {
        this.start();
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      cancelAnimationFrame(this._animationFrame);
    }
  }, {
    key: 'start',
    value: function start() {
      this._playAnimation();
    }
  }, {
    key: 'stop',
    value: function stop() {
      cancelAnimationFrame(this._animationFrame);
    }
  }, {
    key: 'reset',
    value: function reset() {
      this.setState({ frame: 0 });
    }
  }, {
    key: 'render',
    value: function render() {
      var frame = this._getFrame();
      return _react2.default.createElement(
        'div',
        null,
        frame
      );
    }
  }]);
  return SequenceAnimator;
}(_react.Component), _class.displayName = 'SequenceAnimator', _class.propTypes = {
  autoplay: _propTypes2.default.bool,
  duration: _propTypes2.default.number,
  loop: _propTypes2.default.bool,
  easing: _propTypes2.default.oneOf((0, _keys2.default)(Easings)),
  onSequenceEnd: _propTypes2.default.func,
  onAnimationStop: _propTypes2.default.func,
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node])
}, _class.defaultProps = {
  duration: 1000,
  autoplay: true,
  easing: 'linear',
  loop: true,
  children: [],
  onSequenceEnd: function onSequenceEnd() {},
  onAnimationStop: function onAnimationStop() {}
}, _temp);
exports.default = SequenceAnimator;
