// react-sequence-animator/dist/src/SpriteAnimator/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _easingUtils = require('easing-utils');

var Easings = _interopRequireWildcard(_easingUtils);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _SpriteAnimator = require('./SpriteAnimator.scss');

var _SpriteAnimator2 = _interopRequireDefault(_SpriteAnimator);

var _SequenceAnimator = require('../SequenceAnimator');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULT_POSITION = { top: 0, left: 0, width: '100%', height: '100%' };

var SpriteAnimator = (_temp = _class = function (_React$Component) {
  (0, _inherits3.default)(SpriteAnimator, _React$Component);

  function SpriteAnimator() {
    (0, _classCallCheck3.default)(this, SpriteAnimator);

    var _this = (0, _possibleConstructorReturn3.default)(this, (SpriteAnimator.__proto__ || (0, _getPrototypeOf2.default)(SpriteAnimator)).call(this));

    _this._playAnimation = function () {
      _this._animationFrame = requestAnimationFrame(_this._onAnimate);
    };

    _this._onAnimate = function (timestamp) {
      var _this$props = _this.props,
          numOfFrames = _this$props.numOfFrames,
          loop = _this$props.loop,
          easing = _this$props.easing,
          duration = _this$props.duration,
          onSequenceEnd = _this$props.onSequenceEnd,
          onAnimationStop = _this$props.onAnimationStop;


      if (!_this._animationStart) {
        _this._animationStart = timestamp;
      }

      var nextFrame = Math.floor((0, _SequenceAnimator.ease)(easing)(timestamp - _this._animationStart, 0, numOfFrames, duration));

      if (nextFrame > numOfFrames - 1) {
        onSequenceEnd();

        if (loop) {
          nextFrame %= numOfFrames;
          _this._animationStart = timestamp;
        } else {
          nextFrame = -1;
        }
      }

      if (nextFrame > -1) {
        _this.setState({ frame: nextFrame }, function () {
          _this._playAnimation();
        });
      } else {
        onAnimationStop();
      }
    };

    _this.state = {
      frame: 0
    };
    return _this;
  }

  (0, _createClass3.default)(SpriteAnimator, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var autoplay = this.props.autoplay;


      if (autoplay) {
        this.start();
      }
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.numOfFrames !== this.props.numOfFrames) {
        this.start();
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      cancelAnimationFrame(this._animationFrame);
    }
  }, {
    key: 'start',
    value: function start() {
      this._animationStart = null;
      this._playAnimation();
    }
  }, {
    key: 'stop',
    value: function stop() {
      var onAnimationStop = this.props.onAnimationStop;

      cancelAnimationFrame(this._animationFrame);
      onAnimationStop();
    }
  }, {
    key: 'reset',
    value: function reset() {
      this.setState({ frame: 0 });
    }
  }, {
    key: 'render',
    value: function render() {
      var frame = this.state.frame;
      var _props = this.props,
          getPosition = _props.getPosition,
          children = _props.children;

      var _getPosition = getPosition(frame),
          width = _getPosition.width,
          height = _getPosition.height,
          top = _getPosition.top,
          left = _getPosition.left;

      return _react2.default.createElement(
        'div',
        { className: _SpriteAnimator2.default.wrapper, style: { width: width, height: height } },
        _react2.default.createElement(
          'div',
          { className: _SpriteAnimator2.default.innerWrapper, style: { top: -top, left: -left } },
          _react2.default.Children.only(children)
        )
      );
    }
  }]);
  return SpriteAnimator;
}(_react2.default.Component), _class.displayName = 'SpriteAnimator', _class.propTypes = {
  autoplay: _propTypes2.default.bool,
  duration: _propTypes2.default.number,
  loop: _propTypes2.default.bool,
  easing: _propTypes2.default.oneOf((0, _keys2.default)(Easings)),
  children: _propTypes2.default.node,
  getPosition: _propTypes2.default.func.isRequired,
  numOfFrames: _propTypes2.default.number,
  onSequenceEnd: _propTypes2.default.func,
  onAnimationStop: _propTypes2.default.func
}, _class.defaultProps = {
  autoplay: true,
  easing: 'linear',
  loop: true,
  numOfFrames: 0,
  getPosition: function getPosition() {
    return DEFAULT_POSITION;
  },
  onSequenceEnd: function onSequenceEnd() {},
  onAnimationStop: function onAnimationStop() {}
}, _temp);
exports.default = SpriteAnimator;
