// react-sequence-animator/dist/stories/AnimationStory.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AnimationStory = (_temp = _class = function (_React$Component) {
  (0, _inherits3.default)(AnimationStory, _React$Component);

  function AnimationStory(props) {
    (0, _classCallCheck3.default)(this, AnimationStory);

    var _this = (0, _possibleConstructorReturn3.default)(this, (AnimationStory.__proto__ || (0, _getPrototypeOf2.default)(AnimationStory)).call(this, props));

    _this._elemRef = function (el) {
      _this._elem = el;
    };

    _this._onStop = function () {
      var isRunning = _this.state.isRunning;


      if (_this._elem) {
        if (isRunning) {
          _this._elem.stop();
        } else {
          _this._elem.start();
        }
      }

      _this.setState({ isRunning: !isRunning });
    };

    _this._onReset = function () {
      if (_this._elem) {
        _this._elem.reset();
      }
    };

    _this._onDurationChange = function (e) {
      var target = e.target;

      _this.setState({ duration: target.value });
    };

    _this.state = {
      duration: props.duration || 600,
      isRunning: true
    };
    return _this;
  }

  (0, _createClass3.default)(AnimationStory, [{
    key: 'render',
    value: function render() {
      var _state = this.state,
          duration = _state.duration,
          isRunning = _state.isRunning;
      var _props = this.props,
          children = _props.children,
          showReset = _props.showReset;

      var btnText = isRunning ? 'Pause' : 'Play';

      return _react2.default.createElement(
        'div',
        null,
        _react2.default.cloneElement(_react2.default.Children.only(children), { ref: this._elemRef, duration: Number(duration) }),
        _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement(
            'label',
            { htmlFor: 'duration' },
            'Duration:'
          ),
          _react2.default.createElement('input', { type: 'number', value: duration, onChange: this._onDurationChange })
        ),
        _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement('input', { style: { width: '120px' }, type: 'button', value: btnText, onClick: this._onStop })
        ),
        _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement('input', { style: { width: '120px' }, type: 'button', value: 'Reset to start', disabled: !showReset && isRunning, onClick: this._onReset })
        )
      );
    }
  }]);
  return AnimationStory;
}(_react2.default.Component), _class.displayName = 'AnimationStory', _class.propTypes = {
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node]),
  showReset: _propTypes2.default.bool,
  duration: _propTypes2.default.number
}, _temp);
exports.default = AnimationStory;
