// react-sequence-animator/dist/es/src/SpriteAnimator/index.js
import _Object$keys from 'babel-runtime/core-js/object/keys';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _class, _temp;

import React from 'react';
import * as Easings from 'easing-utils';
import PropTypes from 'prop-types';
import styles from './SpriteAnimator.scss';
import { ease } from '../SequenceAnimator';

var DEFAULT_POSITION = { top: 0, left: 0, width: '100%', height: '100%' };

var SpriteAnimator = (_temp = _class = function (_React$Component) {
  _inherits(SpriteAnimator, _React$Component);

  function SpriteAnimator() {
    _classCallCheck(this, SpriteAnimator);

    var _this = _possibleConstructorReturn(this, (SpriteAnimator.__proto__ || _Object$getPrototypeOf(SpriteAnimator)).call(this));

    _this._playAnimation = function () {
      _this._animationFrame = requestAnimationFrame(_this._onAnimate);
    };

    _this._onAnimate = function (timestamp) {
      var _this$props = _this.props,
          numOfFrames = _this$props.numOfFrames,
          loop = _this$props.loop,
          easing = _this$props.easing,
          duration = _this$props.duration,
          onSequenceEnd = _this$props.onSequenceEnd,
          onAnimationStop = _this$props.onAnimationStop;


      if (!_this._animationStart) {
        _this._animationStart = timestamp;
      }

      var nextFrame = Math.floor(ease(easing)(timestamp - _this._animationStart, 0, numOfFrames, duration));

      if (nextFrame > numOfFrames - 1) {
        onSequenceEnd();

        if (loop) {
          nextFrame %= numOfFrames;
          _this._animationStart = timestamp;
        } else {
          nextFrame = -1;
        }
      }

      if (nextFrame > -1) {
        _this.setState({ frame: nextFrame }, function () {
          _this._playAnimation();
        });
      } else {
        onAnimationStop();
      }
    };

    _this.state = {
      frame: 0
    };
    return _this;
  }

  _createClass(SpriteAnimator, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var autoplay = this.props.autoplay;


      if (autoplay) {
        this.start();
      }
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.numOfFrames !== this.props.numOfFrames) {
        this.start();
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      cancelAnimationFrame(this._animationFrame);
    }
  }, {
    key: 'start',
    value: function start() {
      this._animationStart = null;
      this._playAnimation();
    }
  }, {
    key: 'stop',
    value: function stop() {
      var onAnimationStop = this.props.onAnimationStop;

      cancelAnimationFrame(this._animationFrame);
      onAnimationStop();
    }
  }, {
    key: 'reset',
    value: function reset() {
      this.setState({ frame: 0 });
    }
  }, {
    key: 'render',
    value: function render() {
      var frame = this.state.frame;
      var _props = this.props,
          getPosition = _props.getPosition,
          children = _props.children;

      var _getPosition = getPosition(frame),
          width = _getPosition.width,
          height = _getPosition.height,
          top = _getPosition.top,
          left = _getPosition.left;

      return React.createElement(
        'div',
        { className: styles.wrapper, style: { width: width, height: height } },
        React.createElement(
          'div',
          { className: styles.innerWrapper, style: { top: -top, left: -left } },
          React.Children.only(children)
        )
      );
    }
  }]);

  return SpriteAnimator;
}(React.Component), _class.displayName = 'SpriteAnimator', _class.propTypes = {
  autoplay: PropTypes.bool,
  duration: PropTypes.number,
  loop: PropTypes.bool,
  easing: PropTypes.oneOf(_Object$keys(Easings)),
  children: PropTypes.node,
  getPosition: PropTypes.func.isRequired,
  numOfFrames: PropTypes.number,
  onSequenceEnd: PropTypes.func,
  onAnimationStop: PropTypes.func
}, _class.defaultProps = {
  autoplay: true,
  easing: 'linear',
  loop: true,
  numOfFrames: 0,
  getPosition: function getPosition() {
    return DEFAULT_POSITION;
  },
  onSequenceEnd: function onSequenceEnd() {},
  onAnimationStop: function onAnimationStop() {}
}, _temp);
export { SpriteAnimator as default };
